### ISSUE TITLE
How to work with pattern match?

### ISSUE THREAD

**Issue by GalAster** [rust-num/num-bigint#165]
I tried the following methods but none of them worked, how should I use pattern matching?

```rust
fn fibonacci_u(n: BigUint) -> BigUint {
    match n {
        BigUint::from(0) | BigUint::from(1) => n,
        _ => fibonacci_u(n.clone() - BigUint::from(1)) + fibonacci_u(n - BigUint::from(2))
    }
}
```

```rust
fn fibonacci_u(n: BigUint) -> BigUint {
    let zero: BigUint = BigUint::from(0usize);
    let one: BigUint = BigUint::from(1usize);
    let two: BigUint = BigUint::from(2usize);
    match n {
        zero | one => n.clone(),
        _ => fibonacci_u(n.clone() - one) + fibonacci_u(n - two)
    }
}
```

```rust
const ZERO: BigUint = BigUint::from(0usize);
const ONE: BigUint = BigUint::from(1usize);
const TWO: BigUint = BigUint::from(2usize);

fn fibonacci_u(n: BigUint) -> BigUint {
    match n {
        ZERO | ONE => n,
        _ => fibonacci_u(n.clone() - ONE) + fibonacci_u(n - TWO)
    }
}
```

**Comment by cuviper** [comment]
At the language level, pattern matching only works with primitive literal values and structural decomposition. Given that `BigUint` has private fields, and those have even further privacy in `Vec` fields, you can't directly match `BigUint` with values.

You can always use plain `if` conditions like `n == zero || n == one` or `n <= one`. With the `Zero` and `One` traits, this can also be `n.is_zero() || n.is_one()`. If you really want to use patterns, you can do that with an interim conversion to a primitive type using `TryInto` or `ToPrimitive`:

```rust
match n.to_u32() {
    Some(0) | Some(1) => n,
    _ => todo!(),
}
```

**Comment by GalAster** [comment]
This is a good usage example, thanks for your help

**Comment by aid** [comment]
I too came across this issue, where I found the `BigUint` always matched on the first branch of the `match` statement.   As described above, using `if` statements resolved the issue; but I also found that `match` statements with an explicit comparsion worked fine too:

```rust
let v = match n {
    n if n == zero => zero,
    n if n == one => one,
    _ => _fibonacci(&n - &one, cache) + _fibonacci(&n - &two, cache),
};
```

...where `n` is of course a BigUint.

This was a little nicer than a tree of if statements in my particular situation.
